src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,474 @@
+
+/**
+ * <p>
+ * Checks that a local variable is declared and/or assigned, but not used.
+ * Doesn't check
+ * <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30">
+ * pattern variables</a>.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;UnusedLocalVariable&quot;/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * class Test {
+ *
+ *     int a;
+ *
+ *     {
+ *         int k = 12; // violation, assigned and updated but never used
+ *         k++;
+ *     }
+ *
+ *     Test(int a) {   // ok as 'a' is a constructor parameter not a local variable
+ *         this.a = 12;
+ *     }
+ *
+ *     void method(int b) {
+ *         int a = 10;             // violation
+ *         int[] arr = {1, 2, 3};  // violation
+ *         int[] anotherArr = {1}; // ok
+ *         anotherArr[0] = 4;
+ *     }
+ *
+ *     String convertValue(String newValue) {
+ *         String s = newValue.toLowerCase(); // violation
+ *         return newValue.toLowerCase();
+ *     }
+ *
+ *     void read() throws IOException {
+ *         BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+ *         String s; // violation
+ *         while ((s = reader.readLine()) != null) {
+ *         }
+ *         try (BufferedReader reader1 // ok as 'reader1' is a resource and resources are closed
+ *                                     // at the end of the statement
+ *             = new BufferedReader(new FileReader("abc.txt"))) {
+ *         }
+ *         try {
+ *         } catch (Exception e) {     // ok as e is an exception parameter
+ *         }
+ *     }
+ *
+ *     void loops() {
+ *         int j = 12;
+ *         for (int i = 0; j &lt; 11; i++) { // violation, unused local variable 'i'.
+ *         }
+ *         for (int p = 0; j &lt; 11; p++)   // ok
+ *             p /= 2;
+ *     }
+ *
+ *     void lambdas() {
+ *         Predicate&lt;String&gt; obj = (String str) -&gt; { // ok as 'str' is a lambda parameter
+ *             return true;
+ *         };
+ *         obj.test("test");
+ *     }
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code unused.local.var}
+ * </li>
+ * </ul>
+ *
+ * @since 9.3
+ */
+@FileStatefulCheck
+public class UnusedLocalVariableCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_UNUSED_LOCAL_VARIABLE = "unused.local.var";
+
+    /**
+     * An array of increment and decrement tokens.
+     */
+    private static final int[] INCREMENT_AND_DECREMENT_TOKENS = {
+        TokenTypes.POST_INC,
+        TokenTypes.POST_DEC,
+        TokenTypes.INC,
+        TokenTypes.DEC,
+    };
+
+    /**
+     * An array of scope tokens.
+     */
+    private static final int[] SCOPES = {
```

Comment:
why isn't LITERAL_DO needed here?

Reply:
Actually at present `LITERAL_WHILE` and `LITERAL_IF` are also not needed, these token were left by mistake when I removed the partial support for pattern variables.
```
Object o = 12;
if(o instanceof String s) {} // scope is LITERAL_IF
while(o instanceof String s) {} // scope is LITERAL_WHILE
```
But LITERAL_DO can't be the scope as pattern variables introduced in while statements of do-while loops are not accessible in `do {}`
```
Object o = 12;
do {
     System.out.println(s); // s is not accessible here
                            // so scope of 's' would be LITERAL_WHILE
} while (o instanceof String s);
```

Reply:
I am confused then. I was not talking about LITERAL_DO from the prospect of pattern variables. The variable named scopes, so I thought this was a list of scopes to denote where the variable "lives" so it can be used for it's tracking of unused-ness.

So why are only these specific scopes needed?
Does the following type of code not cause an issue with scope in this check?
````
while (condition) {
   int var = 0; // unused
}
int var = 1; // used in different scope
System.out.println(var);
````

Reply:
> So why are only these specific scopes needed?

We need `SLIST` for `{`, `LITERAL_FOR` for `for(int i=0;;);` and 
`OBJBLOCK` for instance var under `{`

> Does the following type of code not cause an issue with scope in this check?

Here the scope will not be LITERAL_WHILE, it will be SLIST(`{`), and `int var = 0;` will be popped out of the scope when we leave SLIST(`{`) so there won't be an issue.